Use histogram for camera sensor to capture significant brightness

pdf opened the pull request:
This is a first pass to fix #25

Results so far are very encouraging.

pdf committed:
Use histogram for camera sensor to capture significant brightness

pdf commented on src/modules/sensors/camera.c:
```diff
@@ -8,6 +8,7 @@
 #define CAMERA_NAME                 "Camera"
 #define CAMERA_ILL_MAX              255
 #define CAMERA_SUBSYSTEM            "video4linux"
+#define HISTOGRAM_STEPS             20
```
Using 20 steps here gives us about 5% slices to average, I tested some other values and they all produce decent results, tweaking this just adjusts granularity.

FedeDP replied:
I'd make HISTOGRAM_STEPS overridable at compile time from CMakeLists.txt.

pdf replied:
I've made the other changes you suggested locally, but I'm afraid I have no idea how to drive CMake.

FedeDP replied:
That is a no-issue. I can do that myself, or I can guide you if you wish!

pdf commented on src/modules/sensors/camera.c:
```diff
@@ -290,18 +309,55 @@ static void recv_frame(int i) {
+    brightness = state.hist.sum[HISTOGRAM_STEPS-1] / state.hist.count[HISTOGRAM_STEPS-1];
+    // Find the highest brightness bucket that has a significant sample count and return the average brightness for that bucket
+    for (int i = HISTOGRAM_STEPS-1; i >= 0; i--) {
+        if (state.hist.count[i] > step_size) {
```
Here we filter out buckets at the top that contain a low number of samples - I had some measurements where a point source of light reflecting off something would produce a handful of pixels at really high brightness. Using `step_size` here is entirely arbitrary - this could also be a fixed number, or some other scale, but using the step size seems to make the filter scale moderately well based on the granularity of the histogram.

FedeDP replied:
What about using a different filter, ie:
1) buckets with a good nice of samples
2) bucket with lowest variance

Second point should give us a bucket with homogeneous values that should produce better results.

pdf replied:
Part of the problem we have is that trying to determine ambient brightness from an image is simply not possible without knowing exactly how a particular camera sensor functions at a specific exposure, and even then I suspect it's error-prone.

1. I think you mean taking the highest population bucket? The problem is that images with large dark areas (due to furniture or similar) still won't return good results.
2. A bucket with low variance _and_ a low sample count is probably not representative though, and by their nature, there's a limit on possible variance imposed by the width of a bucket (which varies based on the total range of the scene).  The scenario I saw in testing, that prompted me to add a filter, was a bucket at the top that contained few samples, that were all quite close to each other, but caused by a small source of bright light, far in excess of anything else in the scene (there were a number of empty buckets below it, before the next substantial count).

Since we have the histogram now, we can probably roughly calculate the interquartile range and use that to eliminate outliers, which is the common statistical method - I'll try to take a look at that, and taking some test measurements, over the weekend.

We'll still probably want to do some filtering of low-pop buckets though, I'm just not sure if there's a better threshold.

FedeDP replied:
Actually I meant using both your current filter AND the second one: first found bucket that satisfies both of them.
This way we should use the highest bucket with a minimum number of samples and with lowest variance.

Obviously there is no "correct" approach, and i really like current algorithm's simpleness. That is why i proposed a small improvement on it :)

pdf replied:
I don't think variance is a good metric for filtering, because it's entirely random how much variance a bucket may contain, depending on where the bucket start/end happens to fall.

I've gone ahead and added some approximate calculations of the interquartile range, and used the general stats model of 1.5*IQR to filter any outlying buckets from the top and bottom of the histogram - this should provide pretty solid results, but I need to do some testing to make sure everything looks good now.

FedeDP replied:
That's great, thanks!

pdf commented:
As I mentioned, it's been quite some time since I wrote any C, so feel free to review harshly.

FedeDP commented:
> As I mentioned, it's been quite some time since I wrote any C, so feel free to review harshly.

Harsh is an adjective that doesn't really suit me :) 

Overall it seems really good, i have just some (minor) things!
Btw can you post some tests here? (before and after, and maybe with different HISTOGRAM_STEPS)!
Thanks for this PR though, it's a great improvement!

FedeDP commented on src/modules/sensors/camera.c:
```diff
@@ -48,6 +55,7 @@ struct state {
     uint32_t pixelformat;
     double *brightness;
     struct buffer buf;
+    struct histogram hist;
```
Why not using 
`struct histogram hist[HISTOGRAM_STEPS]`
instead?
Thus struct histogram would become just:
```
struct histogram {
    double count;
    double sum;
};
```

FedeDP commented on src/modules/sensors/camera.c:
```diff
@@ -201,6 +210,16 @@ static void init(void) {
     state.width = fmt.fmt.pix.width;
     state.height = fmt.fmt.pix.height;
     state.pixelformat = fmt.fmt.pix.pixelformat;
+    state.hist.count = (double*) malloc(HISTOGRAM_STEPS * sizeof(double));
```
No need to malloc here then.

FedeDP commented on src/modules/sensors/camera.c:
```diff
@@ -201,6 +210,16 @@ static void init(void) {
+        perror("Allocating histogram count");
+        return;
+    }
+    state.hist.sum = (double*) malloc(HISTOGRAM_STEPS * sizeof(double));
```
Here neither.

FedeDP commented on src/modules/sensors/camera.c:
```diff
@@ -290,18 +309,55 @@ static void recv_frame(int i) {
 }
 
+static void reset_histogram(void) {
+    state.hist.count = memset(state.hist.count, 0, HISTOGRAM_STEPS * sizeof(double));
```
We  can only memset state.hist here then. (`HISTOGRAM_STEPS * sizeof(struct histogram)`)

FedeDP commented on src/modules/sensors/camera.c:
```diff
@@ -290,18 +309,55 @@ static void recv_frame(int i) {
-    brightness /= state.width * state.height;
+
+    // Default to the average for the highest brightness bucket
+    brightness = state.hist.sum[HISTOGRAM_STEPS-1] / state.hist.count[HISTOGRAM_STEPS-1];
```
Is this really needed though? We parse it anyway!

pdf replied:
I _may_ be able to remove this - I added it as a safety-net when I was testing out the filter mechanism below, in case we didn't match any buckets.

FedeDP replied:
It makes sense, thanks!

FedeDP replied:
But are we sure that
`state.hist.count[HISTOGRAM_STEPS-1] != 0`
in any case?

pdf replied:
Since we've calculated the min/max, the upper bucket is guaranteed to have samples.

FedeDP replied:
You're right, thanks for the clarification!

FedeDP commented on src/modules/sensors/camera.c:
```diff
@@ -312,6 +368,12 @@ static void free_all(void) {
         close(state.device_fd);
     }
+    if (state.hist.count) {
+        free(state.hist.count);
```
No need to free here.

FedeDP commented on src/modules/sensors/camera.c:
```diff
@@ -312,6 +368,12 @@ static void free_all(void) {
+        free(state.hist.count);
+    }
+    if (state.hist.sum) {
+        free(state.hist.sum);
```
No need to free here.

pdf commented:
Will update with the requested changes soon(ish).

> Btw can you post some tests here? (before and after, and maybe with different HISTOGRAM_STEPS)!

Do you have some sort of test process I should apply?

FedeDP commented:
> Do you have some sort of test process I should apply?

I'd say just a couple of captures in a couple of different conditions with both old code and new one, very easily.

pdf commented:
I just remembered I might actually have an ALS sensor somewhere though I have absolutely no idea where it is, probably in one of many boxes in storage.  I might see if I can find that easily - if I can get some comparative measurements that should provide good validation.

If I can't find it, I'll just do as you suggest.

pdf commented:
One problem I'm noticing is that in a dark room, the content of the screen has a significant impact on brightness measurements, since the light output can change dramatically (bright content can actually produce a run-away increase in readings, measurement-on-measurement, as the backlight increases output).  This *might* be affected by auto-exposure compensating for an otherwise dark scene though - I'm going to play with fixed exposure to test that theory, so I can validate that the algo in this PR is viable.

If that doesn't have the desired impact, it might be worth considering having a screen-capture plugin that can compensate for screen content during measurements, though that would mean bringing in display protocol code (probably only X, at least initially).

These changes would be out of scope for this PR, but wanted to put my thoughts down while I have them.

FedeDP commented:
> One problem I'm noticing is that in a dark room, the content of the screen has a significant impact on brightness measurements, since the light output can change dramatically (bright content can actually produce a run-away increase in readings, measurement-on-measurement, as the backlight increases output)

Yes, this is well known. The issue here is that it is quite difficult to correctly compute monitor luminescence, as it depends not only on what's currently displayed (easy on Xorg, i already have some code doing that), but on monitor size and distance from webcam...
But i can see a new exposed method (X only) something like:
`org.clightd.clightd.Backlight GetMonitorBrightness`
or something like that.

Then, in clight, a new config option to support this on X only.

We only need a good algorithm... Have you got any idea?

FedeDP commented:
Note that, while `GetMonitorBrightness` can be a Clightd method, remaining part of algorithm should go directly into Clight.

I think something like: 
* dark ambient brightness -> heavier monitor brightness contribution
* high ambient brightness -> lighter monitor brightness contribution

But, obviously:
* dark ambient brightness -> lower monitor backlight -> thus lighter monitor brightness contribution
* high ambient brightness -> high monitor backlight -> heavier monitor brightness contribution

I think that we can hold an assumption about these 2 amounts' equality, thus just letting clight do something like:
```
monitor_br = clightd.GetMonitorBrightness()
corrected_br = ambient_br - A*monitor_br
```
Where "A" should be our guessed monitor brightness contribution to ambient brightness, something like 1/10, or lower.

pdf commented:
That aligns well with my thoughts. We'd probably need to expose an option to allow users to adjust the base scaling factor (to allow compensation for the backlight strength and distance variables you mention), and we may need to pass the DISPLAY var to Clightd to attach to the right server for capture.

Monitors do typically have a known size and max output (in nits), so we may be able to provide some rough guidance to users willing to look this up on how to set the base backlight scaling, but it's a single value, so it's probably fine leave it up to experimentation (with a vaguely sensible default).

I think it's worth capturing these ideas in a new issue, whether we decide to move on them or not.

FedeDP commented:
> We'd probably need to expose an option to allow users to adjust the base scaling factor 

Yes, and by default it is "0.0" ie: compensation disabled.

> and we may need to pass the DISPLAY var to Clightd to attach to the right server for capture.

Yes, and Xauthority too (it is already done for dpms and gamma).
Adding the method is super easy, hopefully this weekend i'll find the time to commit such a method, then we can move on to clight.

> but it's a single value, so it's probably fine leave it up to experimentation (with a vaguely sensible default).

That's my thought as well!

FedeDP commented:
Oh, and btw i'm really liking your fresh ideas and support, thank you! :)

pdf commented:
You're most welcome, thanks for the project ;-)

FedeDP commented:
@pdf  I just committed b7fc5c8e07022f78c6d98c999752de0176f4e57d with a new "Screen" module to retrieve currently screen-emitted brightness! :) It is X only.

pdf commented:
Awesome! Though I would note that `XGetImage` is quite slow, though that may not be a big deal for very infrequent captures, generally capturing via XCB is much more performant.

FedeDP commented:
> generally capturing via XCB is much more performant.

Yes, but I already rely upon libX11 for everything and did not want to add one more dep! And as you said, captures are not that frequent.
But if you're into it, i'd accept a porting to xcb of various functions (dpms, gamma and screen modules) :)

pdf commented:
> i'd accept a porting to xcb of various functions (dpms, gamma and screen modules)

Maybe if I find myself with an abundance of spare time :wink:

I was thinking though that it might be nice to be able to have screen content measurements occur much more frequently than light measurements (screen output changes frequently but ambient light changes rarely), to allow for more consistent backlight adaptations to screen content.

FedeDP commented:
> I was thinking though that it might be nice to be able to have screen content measurements occur much more frequently than light measurements (screen output changes frequently but ambient light changes rarely), to allow for more consistent backlight adaptations to screen content.

The problem is: how much frequent? As you said, screen content changes way too frequently to be actually sure we are compensating for the right screen-emitted brightness...even if we compute it every 5s (and it would be a huge hit for laptops batteries).

EDIT: i think that when one enables the option, she/he should be well aware of its limitations.

FedeDP commented:
> Maybe if I find myself with an abundance of spare time wink

That's great :)
> The problem is: how much frequent? As you said, screen content changes way too frequently to be actually sure we are compensating for the right screen-emitted brightness...even if we compute it every 5s (and it would be a huge hit for laptops batteries).

Btw note that this is quite easy to be done in Clight.  
I just tend to avoid "useless" complexity, ie: complexity that does not add any real, measurable benefit.

pdf commented:
> The problem is: how much frequent? As you said, screen content changes way too frequently to be actually sure we are compensating for the right screen-emitted brightness...even if we compute it every 5s (and it would be a huge hit for laptops batteries).

Probably configurable, and since we already have power status, a different interval for battery versus AC.  We'd need to do it over XCB though to support this efficiently, happy to punt this idea down the road a bit.

> Btw note that this is quite easy to be done in Clight.

The scheduling almost certainly belongs in Clight, I just raise it here as it relates to the efficiency of the capture.

> I just tend to avoid "useless" complexity, ie: complexity that does not add any real, measurable benefit.

I think there's value to be found here, with a sufficient capture rate.  However, I'm not sure that's feasible from a perf/power budget POV, though there are tradeoffs that can be made between capture rate and capture resolution that might make it possible, since we're only looking for an average brightness. Again though, something that can be investigated later on - for now compensation at capture time is likely a decent compromise.

FedeDP commented:
> Probably configurable, and since we already have power status, a different interval for battery versus AC. 

Ok, i built this idea:
* timeouts[AC_POWER/BATT] -> by default each 30s on AC and 5min on BATT may be, or something similar
* bucket_size -> to compute average (a default value of 5/10 i think it's fine)

>  for now compensation at capture time is likely a decent compromise.

I think it's better to always provide the best solution at once :)

FedeDP commented:
> I think it's better to always provide the best solution at once :)

On a second thought (after some prototypying), i think it is better to leave this feature for a future clight release.
I'd like to port clight to libmodule before (just like I did for clightd), as it would greatly improve codebase and help me with lots of features.

Hopefully i will start the port as soon as i can (this weekend i guess!) in a new branch.
I will track it in a new clight's issue!

EDIT: opened [issue](https://github.com/FedeDP/Clight/issues/84).

pdf commented:
Sounds good to me.

pdf commented:
Sorry @FedeDP I haven't forgotten about this, I just haven't had time to round it out.

FedeDP commented:
No problem! I'm playing with libmodule and Clight's port to it, thus i am not in a rush :)

FedeDP commented:
@pdf  Any news on this? I am currently planning next release around october, and i'd love to include this too!

Thanks :)

pdf commented:
Sounds good, I'll finish up and do a push this weekend - #24 will actually let me do some more accurate measurements, I've been putting it off until I could find time to hack in some fixed values.

pdf committed:
Use (approx) inter-quartile range to remove camera brightness outliers

pdf commented:
Pushed changes and rebased against master, will try to post measurement comparison tomorrow, as I need to capture at different times during the day.

pdf commented:
Start of measurement comparison, more to follow with different ambient environment:

| time   | lighting   | background | master   | histogram |
| ------ | ---------- | ---------- | -------- | --------- |
| midday | natural    | dark       | 0.370235 | 0.853583  |
| midday | natural    | light      | 0.871    | 0.896     |
| night  | none       | dark       | 0.079827 | 0.087     |
| night  | none       | light      | 0.080    | 0.087     |
| night  | artificial | dark       | 0.079827 | 0.158     |
| night  | artificial | light      | 0.096    | 0.121     |

FedeDP commented:
I am currently merging #29 in master. Can you rebase after it? 
Thanks (and obviously thanks for your work!)

pdf commented:
Added some more comparisons - results support the histogram approach being a lot more consistent, and compensating for different scene composition. I can take another test in morning light if I wake up in time, but I'm pretty satisfied as is.

Rebased against master as requested (build is failing due to dependency changes by the looks).

And you're most welcome, sorry for the delay - life has a way of getting in the way of personal code.

FedeDP commented:
> Added some more comparisons - results support the histogram approach being a lot more consistent, and compensating for different scene composition. I can take another test in morning light if I wake up in time, but I'm pretty satisfied as is.

I think they're great, indeed! 

> Rebased against master as requested (build is failing due to dependency changes by the looks).

Yes, i forgot to mention that now clightd master requires libmodule >= 5.0.0 (released today).

I think the most impressive result is:
> midday | natural | dark | 0.370235 | 0.853583

This is a great result: dark background has nearly no effect on captured ambient brightness, it's incredible!

FedeDP commented:
I was testing it out.  
Just a couple of quick thoughts: currently at night here in Italy, indoor with a warm light (i'd expect 0.4-0.5 ambient brightness).
Here are some results:
```
$ busctl call org.clightd.clightd /org/clightd/clightd/Sensor org.clightd.clightd.Sensor Capture "sis" "" 5 ""
sad "/dev/video0" 5 0.574581 0.602687 0.688126 0.692136 0.6274

$ busctl call org.clightd.clightd /org/clightd/clightd/Sensor org.clightd.clightd.Sensor Capture "sis" "" 5 ""
sad "/dev/video0" 5 0.768105 0.646631 0.572012 0.580255 0.560367

$ busctl call org.clightd.clightd /org/clightd/clightd/Sensor org.clightd.clightd.Sensor Capture "sis" "" 5 ""
sad "/dev/video0" 5 0.710122 0.796166 0.821195 0.842465 0.930671
```

I think they're a bit too high values for current indoor brightness.

For contrast, here are 3 results from current master branch:
```
$ busctl call org.clightd.clightd /org/clightd/clightd/Sensor org.clightd.clightd.Sensor Capture "sis" "" 5 ""
sad "/dev/video0" 5 0.365817 0.38551 0.39875 0.407253 0.410423

$ busctl call org.clightd.clightd /org/clightd/clightd/Sensor org.clightd.clightd.Sensor Capture "sis" "" 5 ""
sad "/dev/video0" 5 0.277258 0.276604 0.277449 0.278411 0.278346

$ busctl call org.clightd.clightd /org/clightd/clightd/Sensor org.clightd.clightd.Sensor Capture "sis" "" 5 ""
sad "/dev/video0" 5 0.326498 0.384847 0.414764 0.436742 0.4489
```

pdf commented:
Hrm, that's interesting. This prompted me to go through boxes and dig out my ColorHugALS to try and make some comparitive measurements, but I can't seem to get that to work at all (device always returns 0 for raw read unfortunately), so perhaps my sensor is broken.

I would expect brightness measurements to be slightly higher in general, since we're selecting for the brighter areas of the capture, however in my testing, the difference was within a reasonable margin. I also notice significant variance in your captures - I did all my testing with fixed exposure, can you try that and report results?

FedeDP commented:
> I did all my testing with fixed exposure, can you try that and report results?

Yep, i'll try again tonight!
In the meantime, some more results:
Indoor, morning with only external light. I'd expect around 0.7 ambient brightness.

Patched:
```
sad "/dev/video0" 5 0.68769 0.723179 0.758487 0.759064 0.758543
sad "/dev/video0" 5 0.674613 0.727277 0.754388 0.754103 0.75471
sad "/dev/video0" 5 0.662603 0.71098 0.739372 0.754056 0.75676
```

Current master:
```
sad "/dev/video0" 5 0.412583 0.438822 0.464211 0.465957 0.465207
sad "/dev/video0" 5 0.339469 0.401719 0.440791 0.463239 0.463279
sad "/dev/video0" 5 0.463359 0.463369 0.463322 0.463352 0.463328
```

This looks much better!
Some more results moving my webcam nearer window:

Patched:
```
sad "/dev/video0" 5 0.741176 0.830065 0.87451 0.892717 0.889965
sad "/dev/video0" 5 0.822745 0.886078 0.933333 0.933707 0.931462
sad "/dev/video0" 5 0.93345 0.929636 0.936315 0.93139 0.935741
```

Current:
```
sad "/dev/video0" 5 0.43866 0.456984 0.472278 0.475736 0.473629
sad "/dev/video0" 5 0.432405 0.453693 0.469032 0.469047 0.469051
sad "/dev/video0" 5 0.453256 0.468566 0.469569 0.469672 0.469148
```

Again, it looks a great improvement to me.
My webcam is a cheap USB webcam and it obviously sucks :) See how values jump up and down!

Btw will try again tonight and let you know!

pdf commented:
How are we looking @FedeDP do you need anything more from me?

FedeDP commented:
> How are we looking @FedeDP do you need anything more from me?
 
Hi! I am currently quite busy with real life :)
One more thing i'd like is a comparison between these new results and your ALS device (if possible obviously)! Just for, you know, big data :D

Everything else is looking fine; i am waiting until i got some more time because this is an UX-breaking change (all clight curves should be adjusted) and i'd like to be ready with quick-response for new issues!

pdf commented:
> Hi! I am currently quite busy with real life :)

I know how that goes ;-)

> One more thing i'd like is a comparison between these new results and your ALS device (if possible obviously)!

Unfortunately my ALS device appears to be faulty - I tried various firmwares, and couldn't get it to read non-zero, so I suspect a hardware problem.

FedeDP commented:
> Unfortunately my ALS device appears to be faulty - I tried various firmwares, and couldn't get it to read non-zero, so I suspect a hardware problem.

Very unfortunate :(

I still haven't managed to properly test some more, sorry!

FedeDP commented:
Finally merged! :) Thanks very much for your work!!
